google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriter.java

```diff
@@ -0,0 +1,381 @@
+        builder.batchingSettings,
+        builder.retrySettings,
+        builder.executorProvider,
+        builder.endpoint);
+    this.streamWriter = streamWriterBuilder.build();
+  }
+
+  /**
+   * Writes a JSONArray that contains JSONObjects to the BigQuery table by first converting the JSON
+   * data to protobuf messages, then using StreamWriter's append() to write the data. This will
+   * return a ApiFuture<AppendRowsResponse>, which will be processed by a callback to determine if a
+   * schema update is required.
+   *
+   * @param jsonArr The JSON array that contains JSONObjects to be written
+   * @param allowUnknownFields if true, json data can have fields unknown to the BigQuery table.
+   * @param offset Offset for deduplication
+   * @return ApiFuture<AppendRowsResponse>
+   */
+  public synchronized ApiFuture<AppendRowsResponse> append(
+      JSONArray jsonArr, long offset, boolean allowUnknownFields) {
+    ProtoRows.Builder rowsBuilder = ProtoRows.newBuilder();
+    // Any error in convertJsonToProtoMessage will throw an
+    // IllegalArgumentException/IllegalStateException/NullPointerException and will halt processing
+    // of JSON data.
+    for (int i = 0; i < jsonArr.length(); i++) {
+      JSONObject json = jsonArr.getJSONObject(i);
+      Message protoMessage =
+          JsonToProtoMessage.convertJsonToProtoMessage(this.descriptor, json, allowUnknownFields);
+      rowsBuilder.addSerializedRows(protoMessage.toByteString());
+    }
+    AppendRowsRequest.ProtoData.Builder data = AppendRowsRequest.ProtoData.newBuilder();
+    data.setWriterSchema(ProtoSchemaConverter.convert(this.descriptor));
+    data.setRows(rowsBuilder.build());
+
+    final ApiFuture<AppendRowsResponse> appendResponseFuture =
+        this.streamWriter.append(
+            AppendRowsRequest.newBuilder()
+                .setProtoRows(data.build())
+                .setOffset(Int64Value.of(offset))
+                .build());
+
+    ApiFutures.<AppendRowsResponse>addCallback(
+        appendResponseFuture,
+        new ApiFutureCallback<AppendRowsResponse>() {
+          @Override
+          public void onSuccess(AppendRowsResponse response) {
+            updateDescriptor();
+            LOG.info("AppendRowsResponse success!");
+          }
+
+          @Override
+          public void onFailure(Throwable t) {
+            updateDescriptor();
+            LOG.severe("AppendRowsResponse error: " + t.toString() + ".");
+          }
+        });
+    return appendResponseFuture;
+  }
+
+  /**
+   * Updates the descriptor and BQTable schema. This function is used to support schema updates, and
+   * is called whenever AppendRowsResponse includes an updated schema. The function is synchronized
+   * since it is called through a callback (which is in another thread). If the main thread is
+   * calling append while the callback thread calls refreshAppend(), this might cause some issues.
+   */
+  private synchronized void updateDescriptor() {
```

Comment:
Why not directly pass in a callback to stream writer for it to execute action on schema update? The code could be simpler and there is no need to add callbacks on the response future.

Reply:
Got it.